src/immer.js

```diff
@@ -161,42 +173,61 @@ export class Immer {
     finalizeTree(root, path, patches, inversePatches) {
         const state = root[DRAFT_STATE]
         if (state) {
-            root = this.useProxies
-                ? state.copy
-                : (state.copy = shallowCopy(state.draft))
+            if (!this.useProxies) {
+                state.finalizing = true
+                state.copy = shallowCopy(state.draft, true)
+                state.finalizing = false
+            }
+            root = state.copy
         }
 
         const {onAssign} = this
         const finalizeProperty = (prop, value, parent) => {
-            // Only `root` can be a draft in here.
+            if (value === parent) {
+                throw Error("Immer forbids circular references")
+            }
+
+            // The only possible draft (in the scope of a `finalizeTree` call) is the `root` object.
             const inDraft = !!state && parent === root
 
             if (isDraft(value)) {
-                // prettier-ignore
-                parent[prop] = value =
+                value =
                     // Patches are never generated for assigned properties.
                     patches && inDraft && !state.assigned[prop]
-                        ? this.finalize(value, path.concat(prop), patches, inversePatches)
+                        ? this.finalize(value, path.concat(prop), patches, inversePatches) // prettier-ignore
                         : this.finalize(value)
 
-                // Unchanged drafts are ignored.
+                // Preserve non-enumerable properties.
+                if (Array.isArray(parent) || isEnumerable(parent, prop)) {
+                    parent[prop] = value
+                } else {
+                    Object.defineProperty(parent, prop, {value})
+                }
+
+                // Unchanged drafts are never passed to the `onAssign` hook.
                 if (inDraft && value === state.base[prop]) return
             }
             // Unchanged draft properties are ignored.
             else if (inDraft && is(value, state.base[prop])) {
                 return
             }
-            // Search new objects for unfinalized drafts. Frozen objects should never contain drafts.
-            else if (isDraftable(value) && !Object.isFrozen(value)) {
-                each(value, finalizeProperty)
+            // Search new objects for unfinalized drafts.
+            // Frozen objects should never contain drafts.
+            else if (
+                isDraftable(value) &&
+                !Object.isFrozen(value) &&
+                this.visited.indexOf(value) < 0
```

Comment:
The idea was to be more resilient against accidental reference duplication in the state tree. When an object appears twice in the state tree, it can lead to infinite recursion (which may be perfectly fine since graphs are not supported). I think it's best to make this check opt-in and throw an error instead of silently skipping, but I'll do that in a future PR instead. 👍